File: lib/services/ads_service.dart
import 'dart:async';
import 'package:flutter_unity_ads/flutter_unity_ads.dart'; // Unity Ads
import 'package:google_mobile_ads/google_mobile_ads.dart'; // AdMob
import 'package:ziberlive/providers/app_state_provider.dart'; // For granting rewards
import 'package:ziberlive/config.dart'; // For reward constants

// Define a generic AdReward item (can be kept or removed if AppStateProvider handles all rewards)
class AdReward {
  final String adNetwork;
  final num amount; // Generic amount, specific points determined by constants
  final String type; // Generic type, specific points determined by constants

  AdReward({required this.adNetwork, required this.amount, required this.type});
}

// Enum to help manage ad state for different providers
enum AdLoadState { notLoaded, loading, loaded, failed }

enum AdNetwork { adMob, unityAds }

class AdsService {
  BannerAd? bannerAd;
  InterstitialAd? _interstitialAd;
  RewardedAd? _admobRewardedAd;
  AdLoadState _unityRewardedAdState = AdLoadState.notLoaded;
  AdNetwork? _lastShownRewardedNetwork; // For alternating logic

  // bool isPremium = false; // Replaced by a getter accessing AppStateProvider
  AppStateProvider? _appStateProvider; // To grant rewards and get premium status

  // Getter for premium status, dynamically fetched from AppStateProvider
  bool get isPremium {
    return _appStateProvider?.isPremium ?? false;
  }

  // Method to set the AppStateProvider instance
  void setAppStateProvider(AppStateProvider provider) {
    _appStateProvider = provider;
  }

  // AdMob Ad Unit IDs - Replace with your actual IDs
  static const String _admobBannerAdUnitId = 'ca-app-pub-3940256099942544/6300978111'; // Test ID
  static const String _admobInterstitialAdUnitId = 'ca-app-pub-3940256099942544/1033173712'; // Test ID
  static const String _admobRewardedAdUnitId = 'ca-app-pub-3940256099942544/5224354917'; // Test ID

  // Unity Ads Game ID and Placement IDs - Replace with your actual IDs
  static const String _unityGameIdAndroid = 'YOUR_UNITY_GAME_ID_ANDROID';
  static const String _unityGameIdIOS = 'YOUR_UNITY_GAME_ID_IOS';
  static const String _unityRewardedPlacementId = 'rewardedVideo'; // Default placement

  // StreamControllers for rewarded ad events
  final StreamController<AdReward> _onRewardEarnedController = StreamController<AdReward>.broadcast();
  Stream<AdReward> get onRewardEarned => _onRewardEarnedController.stream;

  final StreamController<String> _onRewardedAdLoadedController = StreamController<String>.broadcast();
  Stream<String> get onRewardedAdLoaded => _onRewardedAdLoadedController.stream; // String is ad network name

  final StreamController<String> _onRewardedAdFailedToLoadController = StreamController<String>.broadcast();
  Stream<String> get onRewardedAdFailedToLoad => _onRewardedAdFailedToLoadController.stream; // String is ad network name

  final StreamController<String> _onRewardedAdDismissedController = StreamController<String>.broadcast();
  Stream<String> get onRewardedAdDismissed => _onRewardedAdDismissedController.stream; // String is ad network name


  Future<void> initialize() async {
    if (isPremium) return;

    // Initialize AdMob
    await MobileAds.instance.initialize();
    print("AdsService: AdMob Initialized.");

    // Initialize UnityAds
    // Determine platform for Game ID
    String gameId = _unityGameIdAndroid; // Default to Android
    if (TargetPlatform.iOS == true) { // Needs import 'package:flutter/foundation.dart';
        gameId = _unityGameIdIOS;
    }

    try {
      await UnityAds.init(
        gameId: gameId,
        testMode: true, // Set to false for production
        onComplete: () {
          print('AdsService: UnityAds Initialization Complete.');
          // Pre-load Unity Ad after initialization
          loadUnityRewardedAd();
        },
        onFailed: (error, message) =>
            print('AdsService: UnityAds Initialization Failed: $error $message'),
      );
    } catch (e) {
      print('AdsService: UnityAds Initialization exception: $e');
    }
  }

  // void setPremium(bool value) { // No longer needed, isPremium is a getter
  //   isPremium = value;
  // }

    // Removed the direct assignment to isPremium here. The getter handles it.
    // The logic to dispose/load ads based on premium status change
    // should ideally be triggered by AppStateProvider when its _isPremium changes,
    // or AdsService could listen to AppStateProvider.
    // For now, existing calls to initialize() or load methods when isPremium is false will work.
    // If AppStateProvider.isPremium becomes true, existing ad instances will be disposed by this logic.
    if (this.isPremium) { // Use the getter
      // Dispose ads if user becomes premium
      bannerAd?.dispose();
      bannerAd = null;
      _interstitialAd?.dispose();
      _interstitialAd = null;
      _admobRewardedAd?.dispose();
      _admobRewardedAd = null;
      // UnityAds doesn't have a direct "dispose" for loaded ads in the same way,
      // but we can stop trying to show them.
      _unityRewardedAdState = AdLoadState.notLoaded;
      print("AdsService: Ads disabled due to premium status.");
    } else {
      // User is not premium, try to load ads
      initialize(); // Re-initialize or just load ads
      loadBannerAd();
      loadInterstitialAd();
      loadAdMobRewardedAd();
      loadUnityRewardedAd();
    }
  }

  void loadBannerAd() {
    if (isPremium || bannerAd != null) return;
    print("AdsService: Loading AdMob Banner Ad...");
    bannerAd = BannerAd(
      adUnitId: _admobBannerAdUnitId,
      size: AdSize.banner,
      request: AdRequest(),
      listener: BannerAdListener(
        onAdLoaded: (Ad ad) => print('AdsService: AdMob Banner Ad loaded.'),
        onAdFailedToLoad: (Ad ad, LoadAdError error) {
          print('AdsService: AdMob Banner Ad failed to load: $error');
          ad.dispose();
          bannerAd = null; // Reset on failure
        },
      ),
    )..load();
  }

  void loadInterstitialAd() {
    if (isPremium || _interstitialAd != null) return;
    print("AdsService: Loading AdMob Interstitial Ad...");
    InterstitialAd.load(
      adUnitId: _admobInterstitialAdUnitId,
      request: AdRequest(),
      adLoadCallback: InterstitialAdLoadCallback(
        onAdLoaded: (ad) {
          print('AdsService: AdMob Interstitial Ad loaded.');
          _interstitialAd = ad;
          _interstitialAd!.fullScreenContentCallback = FullScreenContentCallback(
            onAdDismissedFullScreenContent: (ad) {
              ad.dispose();
              _interstitialAd = null;
              loadInterstitialAd(); // Preload next
            },
            onAdFailedToShowFullScreenContent: (ad, error) {
              ad.dispose();
              _interstitialAd = null;
              print('AdsService: AdMob Interstitial failed to show: $error');
              loadInterstitialAd(); // Preload next
            }
          );
        },
        onAdFailedToLoad: (error) {
          print('AdsService: AdMob Interstitial Ad failed to load: $error');
          _interstitialAd = null; // Reset on failure
        }
      ),
    );
  }

  void showInterstitialAd() {
    if (isPremium) {
      print("AdsService: Not showing Interstitial Ad (Premium User).");
      return;
    }
    if (_interstitialAd == null) {
      print("AdsService: Interstitial Ad not loaded yet. Attempting to load.");
      loadInterstitialAd();
      return;
    }
    print("AdsService: Showing AdMob Interstitial Ad.");
    _interstitialAd?.show();
  }

  // --- Rewarded Ads Logic ---

  void loadAdMobRewardedAd() {
    if (isPremium || _admobRewardedAd != null) return;
    print("AdsService: Loading AdMob Rewarded Ad...");
    RewardedAd.load(
      adUnitId: _admobRewardedAdUnitId,
      request: AdRequest(),
      rewardedAdLoadCallback: RewardedAdLoadCallback(
        onAdLoaded: (RewardedAd ad) {
          print('AdsService: AdMob Rewarded Ad loaded.');
          _admobRewardedAd = ad;
          _onRewardedAdLoadedController.add("AdMob");
          _admobRewardedAd!.fullScreenContentCallback = FullScreenContentCallback(
            onAdDismissedFullScreenContent: (RewardedAd ad) {
              print('AdsService: AdMob Rewarded Ad dismissed.');
              _onRewardedAdDismissedController.add("AdMob");
              ad.dispose();
              _admobRewardedAd = null;
              loadAdMobRewardedAd(); // Preload next
            },
            onAdFailedToShowFullScreenContent: (RewardedAd ad, AdError error) {
              print('AdsService: AdMob Rewarded Ad failed to show: $error');
              ad.dispose();
              _admobRewardedAd = null;
              loadAdMobRewardedAd(); // Preload next
            },
          );
        },
        onAdFailedToLoad: (LoadAdError error) {
          print('AdsService: AdMob Rewarded Ad failed to load: $error');
          _admobRewardedAd = null;
          _onRewardedAdFailedToLoadController.add("AdMob");
        },
      ),
    );
  }

  void loadUnityRewardedAd() async {
    if (isPremium || _unityRewardedAdState == AdLoadState.loading || _unityRewardedAdState == AdLoadState.loaded) {
      return;
    }
    print("AdsService: Loading Unity Rewarded Ad (Placement: $_unityRewardedPlacementId)...");
    _unityRewardedAdState = AdLoadState.loading;
    try {
      // The flutter_unity_ads plugin typically uses a listener approach.
      // Loading is often implicit with checking placement state or calling a load method.
      // We'll use `UnityAds.load()` if available, or rely on `isUnityPlacementReady` before showing.
      // For this example, let's assume a direct load call pattern or that init preloads.
      // We need to listen to events from UnityAds.
      // The `flutter_unity_ads` plugin might require setting up listeners globally or when showing.
      // For simplicity, we'll simulate a load and use a flag.

      // This is a common pattern: check if placement is ready.
      // Actual loading might be tied to `UnityAds.load()` or `UnityAds.showVideo()`.
      // The `flutter_unity_ads` documentation should be consulted for the exact API.
      // For now, we'll assume `UnityAds.load` is the way if explicit loading is needed.
      await UnityAds.load(
        placementId: _unityRewardedPlacementId,
        onComplete: (placementId) {
          print('AdsService: Unity Rewarded Ad ($placementId) Loaded.');
          _unityRewardedAdState = AdLoadState.loaded;
          _onRewardedAdLoadedController.add("UnityAds");
        },
        onFailed: (placementId, error, message) {
          print('AdsService: Unity Rewarded Ad ($placementId) Load Failed: $error $message');
          _unityRewardedAdState = AdLoadState.failed;
          _onRewardedAdFailedToLoadController.add("UnityAds");
        },
      );
    } catch (e) {
      print('AdsService: Unity Rewarded Ad load exception: $e');
      _unityRewardedAdState = AdLoadState.failed;
      _onRewardedAdFailedToLoadController.add("UnityAds");
    }
  }

  bool isAdMobRewardedAdReady() {
    return _admobRewardedAd != null;
  }

  bool isUnityRewardedAdReady() {
    // This depends on the plugin's API. Some use `UnityAds.isReady(placementId)`.
    // Or a successful load callback as handled by `_unityRewardedAdState`.
    return _unityRewardedAdState == AdLoadState.loaded;
  }

  void showRewardedAd({String? rewardContext}) { // Added rewardContext
    if (isPremium) {
      print("AdsService: Not showing Rewarded Ad (Premium User).");
      return;
    }
    print("AdsService: Request to show Rewarded Ad (Context: $rewardContext).");

    bool admobReady = isAdMobRewardedAdReady();
    bool unityReady = isUnityRewardedAdReady();

    AdNetwork? networkToShow;

    if (admobReady && unityReady) {
      // Both ready, alternate
      if (_lastShownRewardedNetwork == AdNetwork.adMob) {
        networkToShow = AdNetwork.unityAds;
      } else { // Includes null case (first time) or if last was Unity
        networkToShow = AdNetwork.adMob;
      }
    } else if (admobReady) {
      networkToShow = AdNetwork.adMob;
    } else if (unityReady) {
      networkToShow = AdNetwork.unityAds;
    }

    if (networkToShow == AdNetwork.adMob) {
      print("AdsService: Showing AdMob Rewarded Ad.");
      _admobRewardedAd!.show(onUserEarnedReward: (AdWithoutView ad, RewardItem reward) {
        print('AdsService: AdMob Reward! Amount: ${reward.amount}, Type: ${reward.type}');
        if (rewardContext == "SettingsAd") {
          _appStateProvider?.grantAdRewardPoints(
            coins: kRewardSettingsAdCoins,
            amazonCouponPoints: kRewardSettingsAdAmazonCouponPoints,
            payPalPoints: kRewardSettingsAdPayPalPoints,
          );
          _appStateProvider?.recordSettingsAdWatched(); // Record after successful settings ad
          print("AdsService: Settings Ad rewards granted via AdMob.");
        } else { // Default to Sync Ad rewards
          _appStateProvider?.grantAdRewardPoints(
            trustScorePoints: kRewardTrustScorePoints,
            communityTreePoints: kRewardCommunityTreePoints,
            incomePoolPoints: kRewardIncomePoolPoints,
            amazonCouponPoints: kRewardAmazonCouponPoints, // Note: Sync ads also give Amazon points in this setup
            payPalPoints: kRewardPayPalPoints,             // Note: Sync ads also give PayPal points
          );
          print("AdsService: Sync Ad rewards granted via AdMob.");
        }
        _onRewardEarnedController.add(AdReward(adNetwork: "AdMob", amount: reward.amount, type: reward.type));
      });
      _lastShownRewardedNetwork = AdNetwork.adMob;
    } else if (networkToShow == AdNetwork.unityAds) {
      print("AdsService: Showing UnityAds Rewarded Ad.");
      UnityAds.showVideo(
        placementId: _unityRewardedPlacementId,
        onComplete: (placementId) {
          print('AdsService: UnityAds Rewarded Ad ($placementId) shown and completed.');
          if (rewardContext == "SettingsAd") {
            _appStateProvider?.grantAdRewardPoints(
              coins: kRewardSettingsAdCoins,
              amazonCouponPoints: kRewardSettingsAdAmazonCouponPoints,
              payPalPoints: kRewardSettingsAdPayPalPoints,
            );
            _appStateProvider?.recordSettingsAdWatched(); // Record after successful settings ad
            print("AdsService: Settings Ad rewards granted via UnityAds.");
          } else { // Default to Sync Ad rewards
            _appStateProvider?.grantAdRewardPoints(
              trustScorePoints: kRewardTrustScorePoints,
              communityTreePoints: kRewardCommunityTreePoints,
              incomePoolPoints: kRewardIncomePoolPoints,
              amazonCouponPoints: kRewardAmazonCouponPoints,
              payPalPoints: kRewardPayPalPoints,
            );
            print("AdsService: Sync Ad rewards granted via UnityAds.");
          }
          _onRewardEarnedController.add(AdReward(adNetwork: "UnityAds", amount: 10, type: "default"));

          _unityRewardedAdState = AdLoadState.notLoaded;
          _onRewardedAdDismissedController.add("UnityAds");
          loadUnityRewardedAd();
        },
        onFailed: (placementId, error, message) {
          print('AdsService: UnityAds Rewarded Ad ($placementId) Failed to Show: $error $message');
          _unityRewardedAdState = AdLoadState.notLoaded;
          loadUnityRewardedAd();
        },
        onStart: (placementId) => print('AdsService: UnityAds Rewarded Ad ($placementId) Started.'),
        onClick: (placementId) => print('AdsService: UnityAds Rewarded Ad ($placementId) Clicked.'),
        onSkipped: (placementId) {
          print('AdsService: UnityAds Rewarded Ad ($placementId) Skipped.');
          _onRewardedAdDismissedController.add("UnityAds");
          _unityRewardedAdState = AdLoadState.notLoaded;
          loadUnityRewardedAd();
        }
      );
      _lastShownRewardedNetwork = AdNetwork.unityAds;
    } else {
      print("AdsService: No Rewarded Ad is currently available from any network. Attempting to load both.");
      if (!admobReady) loadAdMobRewardedAd();
      if (!unityReady && _unityRewardedAdState != AdLoadState.loading) { // Check not already loading for Unity
        loadUnityRewardedAd();
      }
    }
  }

  void dispose() {
    print("AdsService: Disposing AdsService resources.");
    bannerAd?.dispose();
    _interstitialAd?.dispose();
    _admobRewardedAd?.dispose();
    _onRewardEarnedController.close();
    _onRewardedAdLoadedController.close();
    _onRewardedAdFailedToLoadController.close();
    _onRewardedAdDismissedController.close();
    // UnityAds usually has a static `UnityAds.destroy()` or similar, but it's often not needed
    // unless you are completely stopping ads for the app session.
    // Individual ad objects are not typically "disposed" in Unity Ads Flutter plugins.
  }
}

File: Assets/Scripts/AdsService.cs
using UnityEngine;
using System;
// Assuming Google Mobile Ads Unity plugin is imported
// using GoogleMobileAds.Api;
// Assuming Unity Ads SDK is integrated
// using UnityEngine.Advertisements;

public class AdsService : MonoBehaviour
{
    // Placeholder Ad Unit IDs - these would be configured in AdMob and Unity Ads dashboards
    private const string AdMobRewardedAdUnitId = "ca-app-pub-YOUR_ADMOB_APP_ID/YOUR_REWARDED_AD_UNIT_ID";
    private const string UnityRewardedAdPlacementId = "rewardedVideo"; // Default placement ID

    // Placeholder for SDK specific objects
    // private RewardedAd adMobRewardedAd;

    public static AdsService Instance { get; private set; }

    public event Action OnRewardedAdLoaded;
    public event Action<string> OnRewardedAdFailedToLoad;
    public event Action<string> OnRewardedAdSkipped; // User skipped or closed the ad
    public event Action<string, int> OnUserEarnedReward; // Ad network, reward amount/type

    private bool isAdMobRewardedAdLoaded = false;
    private bool isUnityRewardedAdLoaded = false;

    void Awake()
    {
        if (Instance == null)
        {
            Instance = this;
            DontDestroyOnLoad(gameObject);
            InitializeAds();
        }
        else
        {
            Destroy(gameObject);
        }
    }

    private void InitializeAds()
    {
        Debug.Log("AdsService: Initializing...");

        // Initialize AdMob (Illustrative)
        // MobileAds.Initialize(initStatus => {
        //     Debug.Log("AdMob Initialized.");
        //     LoadAdMobRewardedAd();
        // });

        // Initialize UnityAds (Illustrative)
        // Advertisement.Initialize("YOUR_UNITY_GAME_ID", testMode: true, enablePerPlacementLoad: true); // Replace with your Game ID
        // Debug.Log("UnityAds Initialized (or will initialize).");
        // LoadUnityRewardedAd(); // Unity Ads loads placements automatically after initialization if configured

        // For now, let's simulate SDK initialization and ad loading
        Invoke(nameof(SimulateAdMobLoadSuccess), 2f);
        Invoke(nameof(SimulateUnityAdsLoadSuccess), 3f);
    }

    #region AdMob Rewarded Ad (Illustrative)
    private void LoadAdMobRewardedAd()
    {
        Debug.Log("AdsService: Loading AdMob Rewarded Ad...");
        // this.adMobRewardedAd = new RewardedAd(AdMobRewardedAdUnitId);

        // // Called when an ad request has successfully loaded.
        // this.adMobRewardedAd.OnAdLoaded += (sender, args) => {
        //     isAdMobRewardedAdLoaded = true;
        //     Debug.Log("AdMob Rewarded Ad Loaded.");
        //     OnRewardedAdLoaded?.Invoke();
        // };
        // // Called when an ad request failed to load.
        // this.adMobRewardedAd.OnAdFailedToLoad += (sender, args) => {
        //     isAdMobRewardedAdLoaded = false;
        //     Debug.Log("AdMob Rewarded Ad failed to load: " + args.LoadAdError.GetMessage());
        //     OnRewardedAdFailedToLoad?.Invoke("AdMob");
        // };
        // // Called when an ad is shown.
        // this.adMobRewardedAd.OnAdOpening += (sender, args) => Debug.Log("AdMob Rewarded Ad Displayed.");
        // // Called when an ad request failed to show.
        // this.adMobRewardedAd.OnAdFailedToShow += (sender, args) => {
        //     Debug.Log("AdMob Rewarded Ad failed to show: " + args.AdError.GetMessage());
        //     OnRewardedAdSkipped?.Invoke("AdMob"); // Or a specific failed to show event
        // };
        // // Called when the user should be rewarded for interacting with the ad.
        // this.adMobRewardedAd.OnUserEarnedReward += (sender, reward) => {
        //     Debug.Log($"AdMob User earned reward: {reward.Amount} {reward.Type}");
        //     OnUserEarnedReward?.Invoke("AdMob", (int)reward.Amount);
        // };
        // // Called when the ad is closed.
        // this.adMobRewardedAd.OnAdClosed += (sender, args) => {
        //     Debug.Log("AdMob Rewarded Ad Closed. Preloading next...");
        //     isAdMobRewardedAdLoaded = false; // Ad is used
        //     LoadAdMobRewardedAd(); // Preload next
        // };

        // AdRequest request = new AdRequest.Builder().Build();
        // this.adMobRewardedAd.LoadAd(request);
    }

    private void ShowAdMobRewardedAd()
    {
        // if (this.adMobRewardedAd != null && this.adMobRewardedAd.IsLoaded())
        // {
        //     this.adMobRewardedAd.Show();
        // }
        // else
        // {
        //     Debug.Log("AdMob Rewarded ad is not ready yet.");
        //     OnRewardedAdFailedToLoad?.Invoke("AdMob"); // Or a specific "not ready" event
        //     LoadAdMobRewardedAd(); // Attempt to reload
        // }
        Debug.Log("Simulating AdMob Rewarded Ad Show");
        OnUserEarnedReward?.Invoke("AdMob", 10); // Simulate reward
        isAdMobRewardedAdLoaded = false;
        Invoke(nameof(SimulateAdMobLoadSuccess), 5f); // Simulate reload
    }
    #endregion

    #region UnityAds Rewarded Ad (Illustrative)
    private void LoadUnityRewardedAd()
    {
        Debug.Log("AdsService: Loading Unity Rewarded Ad (placement)...");
        // Unity Ads loads automatically for a placement if "enablePerPlacementLoad" is true during init
        // Or you can explicitly load:
        // Advertisement.Load(UnityRewardedAdPlacementId, this);

        // Using IUnityAdsLoadListener (new API)
        // Advertisement.Load(UnityRewardedAdPlacementId, new UnityAdsLoadListener(this));
        // Using IUnityAdsShowListener (new API) for show callbacks

        // For simplicity, we'll rely on automatic loading and check status before showing.
        // And simulate load callbacks.
    }

    // Old Unity Ads Listener (IUnityAdsListener - implement this interface on the class)
    /*
    public void OnUnityAdsReady(string placementId)
    {
        if (placementId == UnityRewardedAdPlacementId)
        {
            Debug.Log("Unity Rewarded Ad Loaded (Placement Ready).");
            isUnityRewardedAdLoaded = true;
            OnRewardedAdLoaded?.Invoke();
        }
    }

    public void OnUnityAdsDidError(string message)
    {
        Debug.LogError($"Unity Ads Error: {message}");
        OnRewardedAdFailedToLoad?.Invoke($"UnityAds: {message}");
        isUnityRewardedAdLoaded = false;
    }

    public void OnUnityAdsDidStart(string placementId)
    {
        Debug.Log($"Unity Ad Started: {placementId}");
    }

    public void OnUnityAdsDidFinish(string placementId, ShowResult showResult)
    {
        if (placementId == UnityRewardedAdPlacementId)
        {
            switch (showResult)
            {
                case ShowResult.Finished:
                    Debug.Log("Unity Ad Finished. User should be rewarded.");
                    OnUserEarnedReward?.Invoke("UnityAds", 10); // Example reward
                    break;
                case ShowResult.Skipped:
                    Debug.Log("Unity Ad Skipped.");
                    OnRewardedAdSkipped?.Invoke("UnityAds");
                    break;
                case ShowResult.Failed:
                    Debug.LogError("Unity Ad Failed to Show.");
                    OnRewardedAdSkipped?.Invoke("UnityAds"); // Or a specific failed to show event
                    break;
            }
            isUnityRewardedAdLoaded = false; // Ad is used
            // Unity Ads often auto-reloads, or you might call LoadUnityRewardedAd() here if not using auto-load
        }
    }
    */

    // New Unity Ads Listener (IUnityAdsLoadListener, IUnityAdsShowListener)
    // These would be implemented by a dedicated listener class or this class
    // For IUnityAdsLoadListener
    public void OnUnityAdsAdLoaded(string placementId)
    {
        if (placementId == UnityRewardedAdPlacementId)
        {
            Debug.Log("Unity Rewarded Ad Loaded (New API).");
            isUnityRewardedAdLoaded = true;
            OnRewardedAdLoaded?.Invoke();
        }
    }

    public void OnUnityAdsFailedToLoad(string placementId, UnityAdsLoadError error, string message)
    {
        if (placementId == UnityRewardedAdPlacementId)
        {
            Debug.LogError($"Unity Ads Failed to Load: {error} - {message}");
            OnRewardedAdFailedToLoad?.Invoke($"UnityAds: {error}");
            isUnityRewardedAdLoaded = false;
        }
    }

    // For IUnityAdsShowListener
    public void OnUnityAdsShowFailure(string placementId, UnityAdsShowError error, string message)
    {
        if (placementId == UnityRewardedAdPlacementId)
        {
            Debug.LogError($"Unity Ads Show Failure: {error} - {message}");
            OnRewardedAdSkipped?.Invoke("UnityAds"); // Or a specific failed to show event
            isUnityRewardedAdLoaded = false;
        }
    }

    public void OnUnityAdsShowStart(string placementId) => Debug.Log($"Unity Ads Show Start: {placementId}");
    public void OnUnityAdsShowClick(string placementId) => Debug.Log($"Unity Ads Show Click: {placementId}");
    public void OnUnityAdsShowComplete(string placementId, UnityAdsShowCompletionState showCompletionState)
    {
        if (placementId == UnityRewardedAdPlacementId)
        {
            if (showCompletionState == UnityAdsShowCompletionState.COMPLETED)
            {
                Debug.Log("Unity Ad Completed. User should be rewarded.");
                OnUserEarnedReward?.Invoke("UnityAds", 10); // Example reward
            }
            else
            {
                Debug.Log($"Unity Ad Show Not Completed: {showCompletionState}");
                OnRewardedAdSkipped?.Invoke("UnityAds");
            }
            isUnityRewardedAdLoaded = false; // Ad is used
            // Unity Ads often auto-reloads, or you might call LoadUnityRewardedAd() here if not using auto-load
        }
    }


    private void ShowUnityRewardedAd()
    {
        // For old API:
        // if (Advertisement.IsReady(UnityRewardedAdPlacementId))
        // {
        //    Advertisement.Show(UnityRewardedAdPlacementId);
        // }
        // else
        // {
        //    Debug.Log("Unity Rewarded ad is not ready yet.");
        //    OnRewardedAdFailedToLoad?.Invoke("UnityAds");
        //    LoadUnityRewardedAd(); // Attempt to reload
        // }

        // For new API (assuming 'this' implements IUnityAdsShowListener or a delegate is passed)
        // Advertisement.Show(UnityRewardedAdPlacementId, this);

        Debug.Log("Simulating UnityAds Rewarded Ad Show");
        OnUserEarnedReward?.Invoke("UnityAds", 15); // Simulate reward
        isUnityRewardedAdLoaded = false;
        Invoke(nameof(SimulateUnityAdsLoadSuccess), 5f); // Simulate reload
    }
    #endregion

    #region Mediation Logic (Simplified)
    public void ShowRewardedAd()
    {
        Debug.Log("AdsService: Received request to show Rewarded Ad.");
        // This is a very simplified "mediation" or selection logic.
        // Real mediation SDKs handle this based on eCPM waterfalls or bidding.
        // Here, we'll just prefer AdMob if ready, then UnityAds.

        if (isAdMobRewardedAdLoaded) // Higher priority (example)
        {
            Debug.Log("AdsService: Attempting to show AdMob Rewarded Ad.");
            ShowAdMobRewardedAd();
        }
        else if (isUnityRewardedAdLoaded)
        {
            Debug.Log("AdsService: Attempting to show UnityAds Rewarded Ad.");
            ShowUnityRewardedAd();
        }
        else
        {
            Debug.LogWarning("AdsService: No Rewarded Ad is currently available from any network.");
            // Optionally, try to load both again or inform the user.
            OnRewardedAdFailedToLoad?.Invoke("None"); // No ad available
            // Attempt to reload ads
            // LoadAdMobRewardedAd(); // Actual SDK calls
            // LoadUnityRewardedAd(); // Actual SDK calls
            Invoke(nameof(SimulateAdMobLoadSuccess), 2f); // Simulate trying to load again
            Invoke(nameof(SimulateUnityAdsLoadSuccess), 3f);
        }
    }

    public bool IsAnyRewardedAdReady()
    {
        return isAdMobRewardedAdLoaded || isUnityRewardedAdLoaded;
    }
    #endregion

    #region Simulation Methods for Standalone Testing
    private void SimulateAdMobLoadSuccess()
    {
        isAdMobRewardedAdLoaded = true;
        Debug.Log("AdsService (Simulated): AdMob Rewarded Ad Loaded.");
        OnRewardedAdLoaded?.Invoke();
    }

    private void SimulateUnityAdsLoadSuccess()
    {
        isUnityRewardedAdLoaded = true;
        Debug.Log("AdsService (Simulated): UnityAds Rewarded Ad Loaded.");
        OnRewardedAdLoaded?.Invoke();
    }

    // Dummy Unity Ads specific enums/classes for compilation without SDK
    public enum UnityAdsLoadError { UNKNOWN, INITIALIZE_FAILED, INTERNAL_ERROR, INVALID_ARGUMENT, TIMEOUT, NO_FILL }
    public enum UnityAdsShowError { UNKNOWN, NOT_INITIALIZED, NOT_READY, VIDEO_PLAYER_ERROR, INVALID_ARGUMENT, NO_CONNECTION, ALREADY_SHOWING, INTERNAL_ERROR }
    public enum UnityAdsShowCompletionState { COMPLETED, SKIPPED, UNKNOWN }

    #endregion
}
